### DISCUSSION TITLE
Not getting errors, only empty array with Policies

### DISCUSSION THREAD

**Discussion by dukesx** [supabase/supabase#3576] [Questions] [answered]
I am not sure if this is just me or others too but I stop getting errors from database if I enable Row level security for a table.

For example, if I try to delete a row, but I input the column value wrong, it won't tell me that value is wrong or there is database error. All I get is an empty array from {data} object.

If I disable Row level security, then I get all expected errors from database such as one, that I mentioned above.

```
const {data, err} = supabase.from(table).delete().filter({
title: 'uuid' 
}) 

if (err) console.log(err)  
if (data) console.log(data)

//with row level security, output is empty [ ] but without it shows that title cannot have uuid as value, which is correct since title is string
```

Any help?

**Comment by steve-chavez** [comment, marked as answer]
> const {data, err} = supabase.from(table).delete().filter({
title: 'uuid' 
}) 

That is not a valid `filter` call, see the [filter docs](https://supabase.io/docs/reference/javascript/filter). It should be like `filter('title', 'eq', 'uuid')`.

You should really be using the [eq function](https://supabase.io/docs/reference/javascript/eq) in this case though.

> If I disable Row level security, then I get all expected errors from database such as one, that I mentioned above.

Maybe you were not sending any filters and RLS enabled just ignored the DELETE operation, while RLS disabled gave you an error because you where deleting the whole table(maybe a FK constraint error).

I'd suggest to retry with the proper function calls as I mentioned above.

In general, type checking errors happen before permissions errors.

**Reply by dukesx** [reply]
@steve-chavez Thank you so much for returning response, i have tried this with .eq() function as well, but then again, it works but doesn't respond with an error that i am violating RLS. 

```jsx
             <Button
              onClick={async () => {
                const { data, error } = await supabase
                  .from("articles")
                  .delete()
                  .eq("id", "5cb5ac3b-a5d4-4df0-9d5c-bc338d199739");

                if (error) {
                  console.log(error);
                }
              }}
            >
              Delete
            </Button>
```

##### Articles Table Structure 
- id(uuid)
 - title(text) 
 - description(text) 
 - lead_author (uuid - linked through FK to profiles) 

DELETE RULE: if auth.uid() = lead_author
```sql
auth.uid( ) = lead_author
```

- **Situation**: I have logged in as a user who is not lead_author.
- **Expectation**: DONOT delete article and console log error of RLS violation when pressed.
- **Reality**: Row is NOT deleted but error is not thrown.

**Reply by steve-chavez** [reply]
>  it works but doesn't respond with an error that i am violating RLS.

Yes, this is how RLS works - it won't return an error when violating the USING clause(it will return an error when violating the WITH CHECK clause though).

What you can do in this case is to use the [single()](https://supabase.io/docs/reference/javascript/single) function, then if your `delete` doesn't delete any row it will throw an error.

```js
const { data, error } = await supabase
  .from("articles")
  .delete()
  .eq("id", "5cb5ac3b-a5d4-4df0-9d5c-bc338d199739")
  .single();
```

**Reply by dukesx** [reply]
> > it works but doesn't respond with an error that i am violating RLS.
> 
> Yes, this is how RLS works - it won't return an error when violating the USING clause(it will return an error when violating the WITH CHECK clause though).
> 
> What you can do in this case is to use the [single()](https://supabase.io/docs/reference/javascript/single) function, then if your `delete` doesn't delete any row it will throw an error.
> 
> ```js
> const { data, error } = await supabase
>   .from("articles")
>   .delete()
>   .eq("id", "5cb5ac3b-a5d4-4df0-9d5c-bc338d199739")
>   .single();
> ```

Okay so now it does return some error but i still haven't been able to understand, the difference between CHECK and USING ? arent they supposed to do the same thing ?

**Reply by dukesx** [reply]
I can verify that with CHECK , row level security errors are thrown as expected, such as i tried on a table with INSERT. can't delete have a CHECK instead of using ?

**Reply by steve-chavez** [reply]
>  can't delete have a CHECK instead of using ?

No, as per the [postgres docs](https://www.postgresql.org/docs/current/sql-createpolicy.html):

*A DELETE policy cannot have a WITH CHECK expression, as it only applies in cases where records are being deleted from the relation, so that there is no new row to check.*

> the difference between CHECK and USING ? arent they supposed to do the same thing ?

No, WITH CHECK is for new rows and only for INSERT and UPDATE statements.

**Reply by dukesx** [reply]
> > can't delete have a CHECK instead of using ?
> 
> No, as per the [postgres docs](https://www.postgresql.org/docs/current/sql-createpolicy.html):
> 
> _A DELETE policy cannot have a WITH CHECK expression, as it only applies in cases where records are being deleted from the relation, so that there is no new row to check._
> 
> > the difference between CHECK and USING ? aren't they supposed to do the same thing ?
> 
> No, WITH CHECK is for new rows and only for INSERT and UPDATE statements.

Thank you so much @steve-chavez , this clears it all out !  i just hope that supabase can figure out a way to tackle missing error messages for RLS, i know that Postgres itself doesn't do much but if the team can create a consistent debugging experience, i think supabase might just score another high. You guys rock ! and if engineering degree wasn't a requirement, i would work for you guys free of charge.

**Reply by steve-chavez** [reply]
Thanks for the feedback @dukesx. 

> i just hope that supabase can figure out a way to tackle missing error messages for RLS, i know that Postgres itself doesn't do much but if the team can create a consistent debugging experience

I see what you mean. What we could do for now is to document the `delete()` + `single()` combo in our [delete docs](https://supabase.io/docs/reference/javascript/delete). If you're up for it, a PR would be very much appreciated :pray:

**Reply by dukesx** [reply]
I think this point just needs just a bit of update, a bit of more explanation with RLS with updates to example to use select()
```
If you use delete() with filters and you have RLS enabled, only rows visible through SELECT policies are deleted. Note that by default no rows are visible, so you need at least one SELECT/ALL policy that makes the rows visible.
```
